fix: S3 uris are allowed capital letters

CallumNZ opened the pull request.

junghao commented on .github/workflows/reusable-copy-to-s3.yml:
```diff
@@ -81,7 +81,7 @@ jobs:
       - name: validate
         env:
-          REGEXP_S3_BUCKET: ^s3://[a-z0-9_/.-]+$
+          REGEXP_S3_BUCKET: ^s3://[a-zA-Z0-9_/.-]+$
```
Since A-Z being added in, we can simply use `\w` to represent `a-zA-Z0-9_`

Also, not sure how it worked, but the `.` here means everything. And the `/` there doesn't seem to be valid?

I think the regex should be `\w\/\.-` ? Or it's not actually a regex, just something similar?

ozym replied:
By default the end point bucket needs to be lower case, e.g. 

https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html

> The following naming rules apply for directory buckets.
> 
> - Be unique within the chosen AWS Region and Availability Zone.
> - Name must be between 3 (min) and 63 (max) characters long, including the suffix.
> - Consists only of lowercase letters, numbers and hyphens (-).

> - Begin and end with a letter or number.
> - Must include the following suffix: --azid--x-s3.
> - Bucket names must not start with the prefix xn--.

> - Bucket names must not start with the prefix sthree-.

> - Bucket names must not start with the prefix sthree-configurator.
 
> - Bucket names must not start with the prefix amzn-s3-demo-.
> - Bucket names must not end with the suffix -s3alias. This suffix is reserved for access point alias names. For more information, see [Using a bucket-style alias for your S3 bucket access point](https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-points-alias.html).

> - Bucket names must not end with the suffix --ol-s3. This suffix is reserved for Object Lambda Access Point alias names. For more information, see [How to use a bucket-style alias for your S3 bucket Object Lambda Access Point](https://docs.aws.amazon.com/AmazonS3/latest/userguide/olap-use.html#ol-access-points-alias).

> - Bucket names must not end with the suffix .mrap. This suffix is reserved for Multi-Region Access Point names. For more information, see [Rules for naming Amazon S3 Multi-Region Access Points](https://docs.aws.amazon.com/AmazonS3/latest/userguide/multi-region-access-point-naming.html).


For objects:

You can use any UTF-8 character in an object key name. However, using certain characters in key names can cause problems with some applications and protocols. The following guidelines help you maximize compliance with DNS, web-safe characters, XML parsers, and other APIs.

and then there's a list of recommendations:
https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html

wilsonjord replied:
I think the problem is that the workflow input `s3-bucket` in practice can be bucket-only or full s3 URI. For example, from `docker-gamit-base`:
```
fetch:
    uses: GeoNet/Actions/.github/workflows/reusable-copy-to-s3.yml@main
    with:
      artifact-name: gamit
      artifact-path: ./gg/
      s3-bucket: s3://yum-prod.geonet.org.nz/docker-extras/docker-gamit/
      cp-or-sync: cp
      direction: from # 'to' or 'from'
```
the reusable s3 workflow uses `aws s3` as the copy mechanism, so I think that it should be treated as a  URI.

junghao reviewed (commented):
Sorry I just popped in this PR and saw the unusual regex.

CallumNZ commented:
@junghao 

- It seems you don't need to escape `/` is this environment.
- I have added the full set of allowed characters as per AWS documentation: https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
- `.` within `[]` seems to work as a literal.
- I opted not to use `\w` because I think writing them out in full is clearer.
- I have updated the description of the parameter `s3-bucket` to indicate that it is being used as a URI, not simply a bucket name.
- I tried to deprecate the input to then use a clearer name (e.g. `s3-bucket-uri`), but that functionality is only available for Actions, not Workflows. So for now, I think we'll have to live with the misleading input name `s3-bucket`.

junghao commented:
@CallumNZ 
Would  `'` and `!` causing issue when passing the regex string to https://github.com/GeoNet/Actions/blob/main/.github/workflows/reusable-copy-to-s3.yml#L96 ?

CallumNZ commented:
@junghao good catch. A `'` caused issues in script. Enclosing the regex pattern in `""` fixes it.
